### ISSUE TITLE
Improvement!

### ISSUE THREAD

**Issue by Kai764** [EdyJ/blender-to-unity-fbx-exporter#25]
**Thank you very much for this addon, it really makes with importing Blender models into Unity a lot easier!**

While all the individual object's positions and rotations are correct, it seems like I still need to rotate the whole model by 180° on the Y-axis in Unity to make it face in the right direction.

So I tried making a fix for it (the comments in the code should hopefully be self explanatory, although I am not 100% sure about why certain code segments seem to be required).
Maybe there is a better way, but this seems to work in all scenarios I tested (including more than one child object).

This is the code starting from line 130.
I have also attached the full file below (although as .txt, as .py files are not supported for uploading in this case):

`# Reset parent's inverse so we can work with local transform directly
		reset_parent_inverse(ob)

		# Create a copy of the local matrix and set a X-90/Z-180 matrix
		mat_original = ob.matrix_local.copy()		
		
		rot_x = mathutils.Matrix.Rotation(math.radians(-90.0), 4, 'X')
		rot_z = mathutils.Matrix.Rotation(math.radians(-180.0), 4, 'Z')
		
		ob.matrix_local = rot_x @ rot_z	
		
		# Apply the rotation to the object
		apply_rotation(ob)
		
		if (ob.parent):
			# Reapply the previous local transform with an X+90 and Y+180 rotation
			ob.matrix_local = mat_original @ mathutils.Matrix.Rotation(math.radians(90.0), 4, 'X') @ mathutils.Matrix.Rotation(math.radians(180.0), 4, 'Y')
		else:
			# Reapply the previous local transform with only an X+90 rotation
			ob.matrix_local = mat_original @ mathutils.Matrix.Rotation(math.radians(90.0), 4, 'X')
			
			# Correct position by inverting X and Y values
			loc_or = ob.location		
			ob.location = (-loc_or.x, -loc_or.y, loc_or.z)
			
			# Correct rotation by inverting Y value
			rot_or = ob.rotation_euler
			ob.rotation_euler = (rot_or.x, -rot_or.y, rot_or.z)`

**Hope this helps and of course please let me know if you have any suggestions or if there might be a better way!**

[blender-to-unity-fbx-exporter.txt](https://github.com/EdyJ/blender-to-unity-fbx-exporter/files/8995824/blender-to-unity-fbx-exporter.txt)

**Comment by EdyJ** [comment]
Thank you for your contribution, but this seems a specific scenario applicable to your case only.

In Blender, the "Front" direction is -Y (object facing the negative Y direction). Surely your objects are facing the positive Y direction when all their rotations are 0. This means their natural orientation is facing "Backwards". They conserve this orientation when imported in Unity, so you see that they need an additional rotation of 180 degrees. It's not a fault in the exporter.

The solution is adhering to the Blender convention for the "Front" direction, which is -Y (negative Y). This fbx exporter assumes this convention, so trying to fix any possible corner cases outside it is out of the scope of the add-on.

**Comment by Kai764** [comment]
Thank you for your response.

I know this is a bit of a hackery, compared to the simple 90° X-axis fix.
But every model I imported with this addon needed to be rotated by 180° to get it facing in the right direction.

But do you mean I am using Blender's coordinate system wrong?
This is how my objects face in Blender:

![Blender Rotation](https://user-images.githubusercontent.com/53895707/176119289-7be88a5c-59d1-467e-aa8b-1afe30e14a5c.png)

So should I actually always have my models facing the opposite direction inside Blender (regardless of whether I want to export them to Unity)?
To me, the way I have it currently seems more intuitive, but of course if it's just wrong, I will change my models.
Then your original rotation script should work perfectly as well without needing to do any changes of course.

Let me know!

**Comment by EdyJ** [comment]
Exactly, your model is actually facing backwards in Blender. This is the correct orientation for a cone pointing in the _Front_ direction:

![image](https://user-images.githubusercontent.com/6773528/176133605-4a071963-0c20-45e2-b00e-3f849d9b51e6.png)

When this model is exported to FBX and imported into Unity, the -Y direction will be translated to the +Z direction, which is the _Front_ direction in Unity.

You can verify the _Front_ direction in Blender by pressing Numpad-1, which brings the "Front Orthographic" view. In this view you must be seeing the front of your model, in this case the tip of the cone:

![image](https://user-images.githubusercontent.com/6773528/176134618-c5199596-705b-4869-b6da-7455a78cb103.png)

**Comment by Kai764** [comment]
Thank you for the heads up, I will then change the orientation of my models, which would make all my changes to the script redundant of course.
With all the swapped axes compared to Unity, it still seems a bit weird to me that +y means backwards and +x means left, but I guess I have to get used to it...

**Comment by EdyJ** [comment]
Pretty much every 3D software out here arbitrarily chooses their coordinate system and the meaning of each axis. Here's an (incomplete) example:
![image](https://user-images.githubusercontent.com/6773528/176151111-45570159-e480-4135-ba43-87f08b6d771b.png) ![image](https://user-images.githubusercontent.com/6773528/176151172-5b4886e3-881b-4ba4-af82-990654e45499.png)

**Comment by Kai764** [comment]
Thank you for this graph, I didn't know the coordinate systems were that arbitrary...

Personally, Z forward, Y up and X right makes the most sense to me, like Unity or DirectX have it.
Unreal also seems good enugh to me, just with Y forward and Z up.
But for the rest it looks a bit intuitive to be honest.

I guess you have to make the best out of it, and at least your addon will work without any problems then.
